### ISSUE TITLE
Support for loading other scenarios in ember --serve

### ISSUE THREAD

**Issue by samselikoff** [samselikoff/ember-cli-mirage#1427]
![image](https://user-images.githubusercontent.com/2922250/48871536-245fff00-edb3-11e8-9f1c-e0ab6a3d0a82.png)

---

2 upvotes from Canny

**Comment by teejayhh** [comment]
I am building an addon inhouse atm which does exactly that. Its fairly straightforward to implement it yourself. Without much hacking you can modify the default scenario to load other scenarios. You can create a scenario manager  for instance, which either uses a url parameter or local storage to retrieve a scenario name. In the default scenario you can use the `this` or `server._config.scenarios` var to get a list of all available scenarios inside of your scenario folder. Now what's left is to match up the scenario from the list with your url param value or whatever. This approach has been working out pretty well for us. Give it a try.

**Comment by allthesignals** [comment]
The use case for us is that we need to user test our apps under specific user scenarios. For example, user has xyz assigned projects and must complete abc tasks. I'll give @teejayhh's solution a try!

**Comment by samselikoff** [comment]
Yeah I've come to believe this feature isn't really something Mirage needs to have a first-class API to support. It's basically a subset of "run different code under certain conditions" and those conditions could be build-time, run-time or something else.

I think a future API that might make sense for Mirage is a sort of "reset". I mean today you can do this with `server.db.emptyData()`, and then you can run a scenario through it. So yeah I think I'll be holding off on an official solution/recommendation here for a bit.

**Comment by samselikoff** [comment]
Some examples on the new site on how people are solving this would be great though!

**Comment by allthesignals** [comment]
@samselikoff hmmm that surprises me a little bit because the "scenario" concept in Mirage seems first-class through its usage in the folder hierarchy: a file entry in "mirage/scenarios/" implies the ability to add more "scenarios", as you would factories or fixtures or whatever else. On instinct I went around looking for a way to switch out the scenarios environmentally or through some other means.

I do think updating the docs with some examples will help clear up any confusion—I'm happy to look into that!—but I'm not sure it makes sense to dismiss this just yet.

Thanks!

**Comment by samselikoff** [comment]
The other thing is, in `miragejs` we're calling them `seeds()` instead of `scenarios: { default() }` for precisely this reason. (No other scenario support was ever implemented.)

**Comment by teejayhh** [comment]
I feel like it would not take too much to expose a simple api. For the project I was working I needed 4 key peaces. 
1. Make scenarios useful, other then a function which creates seed data
  - Instead of a function this is now an object with options, name, descriptions, endpoints, custom responses 
2. The default scenario is the entry point to load a different scenario. In my case I have created a ScenarioLoader singleton which uses local storage to determine which scenario to pick. No one likes url parameters.
3. The mirage config loads a MirageServerProxy  which augments/taps the put/post/del/get functions and   returns something like this.
```js
return mirageServer.post(route, this._createCallback(this.scenario, route, data, 'post'), options)
```

The _createCallback needs to return a function  which returns 
```js
     // create a new RouteHandler to create a normal proper mirage response
      const routeHandler = new RouteHandler({
        schema,
        verb: requestType,
        rawHandler: data,  // modify if you like for some cool use cases
        path,
        serializerOrRegistry: server.serializerOrRegistry,
      })

      return routeHandler.handle(request).then((mirageResponse) => {
        return new Response(...mirageResponse)
      })
```
4. And as the fourth part you probably need a component which lets you pick a scenario.

I hope this helps @allthesignals

**Comment by allthesignals** [comment]
Docs PR here: https://github.com/samselikoff/ember-cli-mirage/pull/1802
